pkg/karmadactl/promote.go

```diff
@@ -0,0 +1,479 @@
+	secret, err := clusterClient.CoreV1().Secrets(o.ClusterNamespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	clusterInfos[name].BearerToken = string(secret.Data[clusterv1alpha1.SecretTokenKey])
+	clusterInfos[name].CAData = string(secret.Data[clusterv1alpha1.SecretCADataKey])
+	return nil
+}
+
+// printObjectAndPolicy print the converted resource
+func printObjectAndPolicy(obj *unstructured.Unstructured, gvr schema.GroupVersionResource, opts CommandPromoteOption) error {
+	printer, err := opts.Printer(nil, nil, false, false)
+	if err != nil {
+		return fmt.Errorf("failed to initialize k8s printer. err: %v", err)
+	}
+
+	if err = printer.PrintObj(obj, os.Stdout); err != nil {
+		return fmt.Errorf("failed to print the resource template. err: %v", err)
+	}
+
+	if len(obj.GetNamespace()) == 0 {
+		cpp := buildClusterPropagationPolicy(gvr, opts)
+		if err = printer.PrintObj(cpp, os.Stdout); err != nil {
+			return fmt.Errorf("failed to print the ClusterPropagationPolicy. err: %v", err)
+		}
+	} else {
+		pp := buildPropagationPolicy(gvr, opts)
+		if err = printer.PrintObj(pp, os.Stdout); err != nil {
+			return fmt.Errorf("failed to print the PropagationPolicy. err: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// preprocessResource delete redundant fields to convert resource as template
+func preprocessResource(obj *unstructured.Unstructured) error {
+	// remove fields that generated by kube-apiserver and no need(or can't) propagate to member clusters.
+	if err := prune.RemoveIrrelevantField(obj); err != nil {
+		return err
+	}
+
+	addOverwriteAnnotation(obj)
+
+	return nil
+}
+
+// addOverwriteAnnotation add annotation "work.karmada.io/conflict-resolution" to the resource
+func addOverwriteAnnotation(obj *unstructured.Unstructured) {
+	annotations := obj.DeepCopy().GetAnnotations()
+	if _, exist := annotations[workv1alpha2.ResourceConflictResolutionAnnotation]; !exist {
+		annotations[workv1alpha2.ResourceConflictResolutionAnnotation] = workv1alpha2.ResourceConflictResolutionOverwrite
+		obj.SetAnnotations(annotations)
+	}
+}
+
+// createOrUpdatePropagationPolicy create PropagationPolicy in karmada control plane
+func createOrUpdatePropagationPolicy(karmadaClient *karmadaclientset.Clientset, gvr schema.GroupVersionResource, opts CommandPromoteOption) error {
+	name := opts.name + "-propagation"
+
+	_, err := karmadaClient.PolicyV1alpha1().PropagationPolicies(opts.Namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil && apierrors.IsNotFound(err) {
+		pp := buildPropagationPolicy(gvr, opts)
+		_, err = karmadaClient.PolicyV1alpha1().PropagationPolicies(opts.Namespace).Create(context.TODO(), pp, metav1.CreateOptions{})
+
+		return err
+	}
+	if err != nil {
+		return fmt.Errorf("failed to get PropagationPolicy(%s/%s) in control plane: %v", opts.Namespace, name, err)
+	}
+
+	// PropagationPolicy already exists, not to create it
+	// TODO: support to auto edit PropagationPolicy if it already exists
```

Comment:
I don't suppose we gonna support that dangerous operation.

Reply:
Same as the ClusterPropagationPolicy.

Reply:
Thanks for your reviews. I will do it.